src/main/java/com/frostwire/jlibtorrent/TorrentStats.java

```diff
@@ -0,0 +1,150 @@
+package com.frostwire.jlibtorrent;
+
+import com.frostwire.jlibtorrent.alerts.Alert;
+import com.frostwire.jlibtorrent.alerts.AlertType;
+import com.frostwire.jlibtorrent.alerts.TorrentAlert;
+
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.Queue;
+
+/**
+ * It will start an alertEvent listener to keep track of the download
+ * and upload Speed (bytes/sec). The granularity selected will depend
+ * on the speed in which the results of the speed of the Alert event
+ * listener is retrieved, which can have a delay in milliseconds.
+ *
+ * @author haperlot
+ */
+
+public final class TorrentStats {
+
+    private final SessionManager sessionManager;
+    private final TorrentHandle torrentHandle;
+    private final long samplingIntervalInMs;
+    private final int[] STATS = {AlertType.STATS.swig()};
+    private final long MAX_SAMPLES;
+    private final Queue<Integer> downloadRate = new LinkedList<>();
+    private final Queue<Integer> uploadRate = new LinkedList<>();
+    private long tStart; //for collecting sampling interval time
+
+    public enum Metric {
+        UploadRate,
+        DownloadRate
+    }
+
+    public TorrentStats(final SessionManager sessionManager, TorrentHandle torrentHandle, long samplingIntervalInMs, long maxHistoryInMs) {
+        this.sessionManager = sessionManager;
+        this.torrentHandle = torrentHandle;
+        this.samplingIntervalInMs = samplingIntervalInMs;
+        this.tStart = System.currentTimeMillis();
+        this.MAX_SAMPLES = maxHistoryInMs / samplingIntervalInMs;
+        startAlertListener();
+    }
+
+    /**
+     * It will start the alert event listener, as soon as the sampling interval
+     * meets the requested time it will save the values on queues. If the queues
+     * reach max size denoted by the maxHistory time, they will poll() the head
+     * element, in our case the oldest inserted.
+     */
+    public void startAlertListener() {
+        sessionManager.addListener(new AlertListener() {
+            @Override
+            public int[] types() {
+                return STATS;
+            }
+
+            @Override
+            public void alert(Alert<?> alert) {
+
+                if (!(alert instanceof TorrentAlert<?>)) {
+                    return;
+                }
+                //if not eq to the torrentHandle return.
+                if (!((TorrentAlert<?>) alert).handle().swig().op_eq(torrentHandle.swig())) {
+                    return;
```

Comment:
I wonder if this check is enough...

I believe the `.swig()` is basically the pointer address to the object in C++ land, to be honest, I'm not sure if this down in C++ land might be pointing to a copy of the object. I don't think there's a guarantee that this is the same exact object.

Therefore, I'd try to ask for something that would uniquely identify the `torrentHandle`, what you know for sure, will be the same, will be the torrent `info_hash`, which is a `SHA1` hash, which is a 20-byte array.

We have classes to manipulate and compare these hashes.

Now... perhaps I'm wrong, and the alerts DO have that pointer to the `torrentHandle` instance, I'd have to read the C++ code, and if I'm wrong, this would def. be much faster.

CC: @aldenml